Assets/Mirror/Runtime/NetworkIdentity.cs

```diff
@@ -183,30 +185,133 @@ bool ThisIsASceneObjectWithPrefabParent(out GameObject prefab)
+        //   => BuildPipeline.isBuildingPlayer check solves that
+        void AssignSceneID()
+        {
+            // we only ever assign sceneIds at edit time, never at runtime.
+            // by definition, only the original scene objects should get one.
+            // -> if we assign at runtime then server and client would generate
+            //    different random numbers!
+            if (Application.isPlaying)
+                return;
+
+            // no valid sceneId yet, or duplicate?
+            NetworkIdentity existing;
+            bool duplicate = sceneIds.TryGetValue(m_SceneId, out existing) && existing != null && existing != this;
+            if (m_SceneId == 0 || duplicate)
+            {
+                // if a scene was never opened and we are building it, then a
+                // sceneId would be assigned to build but not saved in editor,
+                // resulting in them getting out of sync.
+                // => don't ever assign temporary ids. they always need to be
+                //    permanent
+                // => throw an exception to cancel the build and let the user
+                //    know how to fix it!
+                if (BuildPipeline.isBuildingPlayer)
+                    throw new Exception("Scene " + gameObject.scene.path + " needs to be opened and resaved before building, because the scene object " + name + " has no valid sceneId yet.");
+
+                // if we generate the sceneId then we MUST be sure to set dirty
+                // in order to save the scene object properly. otherwise it
+                // would be regenerated every time we reopen the scene, and
+                // upgrading would be very difficult.
+                // -> Undo.RecordObject is the new EditorUtility.SetDirty!
+                // -> we need to call it before changing.
+                Undo.RecordObject(this, "Generated SceneId");
+
+                // generate random sceneId
+                // range: 3 bytes to fill 0x00FFFFFF
+                m_SceneId = (uint)UnityEngine.Random.Range(0, 0xFFFFFF);
+                Debug.Log(name + " in scene=" + gameObject.scene.name + " sceneId assigned to: " + m_SceneId.ToString("X") + (duplicate ? " because duplicated" : ""));
+            }
+
+
+            // add to sceneIds dict no matter what
+            // -> even if we didn't generate anything new, because we still need
+            //    existing sceneIds in there to check duplicates
+            sceneIds[m_SceneId] = this;
+        }
+
+        // copy scene build index into sceneId for scene objects.
+        // this is the only way for scene file duplication to not contain
+        // duplicate sceneIds as it seems.
+        // -> sceneId before: 0x00AABBCCDD
+        // -> then we put buildIndex into the 0x00 part
+        // => ONLY USE THIS FROM POSTPROCESSSCENE!
+        public void SetSceneIdSceneIndexByteInternal()
+        {
+            byte buildIndex = (byte)gameObject.scene.buildIndex;
```

Comment:
build index byte is never saved.

if you reorder, it doesn't matter because this is called when building / when pressing play after all reordering is done.

duplicating scene is fine: if you duplicate sceneA with objectA with assetId 0x00AAAAAA then you get sceneB with objectA with assetId 0x00AAAAAA. then when building or when pressing player, OnPostProcessScene copies '0x00' for sceneA into objectA from sceneA, and '0x01' for sceneB into objectA from sceneB.

Reply:
How do you prevent it from being saved? 

You may want to clean up the scene id before assigning it here then.  how about:
```cs
m_SceneId &= 0x00FFFFFF;
```
before you add a buildIndex?

Reply:
OnPostProcessScene is only called when building (no sceneid is set when building, exception is thrown instead if it still needs to be set before building) or when pressing play (changes during play mode are never saved)